### ISSUE TITLE
All models get stuck in "pending" state

### ISSUE THREAD

**Issue by sammcj** [gpustack/gpustack#208]
**Describe the bug**

I wanted to try out GPU Stack so after reading the docs and deploying with docker I've found that no matter what model I deploy - it gets stuck in "pending" and never deploys.

<img width="2431" alt="image" src="https://github.com/user-attachments/assets/3b1b6b54-0ac2-4b86-aa40-ce240167d14f">

<img width="1975" alt="image" src="https://github.com/user-attachments/assets/69751913-cc9c-4ffb-9ab2-17bdf23f0406">

All my many other containerised apps are working fine and there's no errors in the logs - it just never seems to do anything?

**Steps to reproduce**

1. Spin up GPU Stack from docker / docker compose
2. Deploy a model (Ollama or Huggingface)
3. Wait...
4. Nothing happens (forever "pending")

**Result**

Deploying a model only gets as far as pending

**Expected behavior**

Model deployed

**Environment**

- GPUStack version: The official docker image says its "Server 0.0.0, UI 9bd6ded" but I pulled the latest image today (2024-08-30) which has the SHA "sha256:eac62dfc76ee67991af2c29f958da051886f8e1a36d986b6ea3b1dd076bbe3c3" which looks like it was created yesterday: "2024-08-29T21:10:16.077550927Z"
- OS: Fedora 40, Docker 27.2.0, Compose v2.29.2, NV Container toolkit 1.16.1-1
- GPU: 1x 3090, 2x A4000

**Additional context**

```
docker logs -f gpustack
2024-08-29T21:10:17+00:00 - gpustack.server.server - INFO - Starting GPUStack server.
2024-08-29T21:10:17+00:00 - gpustack.server.server - INFO - Running database migration.
2024-08-29T21:10:19+00:00 - gpustack.server.server - INFO - Database migration completed.
2024-08-29T21:10:19+00:00 - gpustack.worker.worker - INFO - Starting GPUStack worker.
2024-08-29T21:10:19+00:00 - gpustack.server.server - INFO - Serving on 0.0.0.0:80.
2024-08-29T21:10:19+00:00 - gpustack.worker.worker - INFO - Serving worker APIs on 0.0.0.0:10150.
2024-08-29T21:10:19+00:00 - gpustack.worker.exporter - INFO - Serving metric exporter on 0.0.0.0:10151.
2024-08-29T21:10:19+00:00 - gpustack.worker.worker_manager - INFO - Registering worker: gpustack
```

**Comment by gitlawr** [comment]
It seems that the worker registration never completes.
1. Can you turn on debug mode to see if there's more useful info?
Ref: https://docs.gpustack.ai/troubleshooting/#configure-log-level
2. Can you check if status of workers & GPUs is expected in the `Resources` page?
3. Can you exec into the container and do `curl -v http://localhost/v1/workers`?

**Comment by sammcj** [comment]
Sorry was away over the weekend! Will set the log level tomorrow and report back :)

**Comment by sammcj** [comment]
<img width="2480" alt="SCR-20240902-hmlc" src="https://github.com/user-attachments/assets/7d46cc9a-b12e-40f8-9767-a4ddcbc59da8">

<img width="2488" alt="SCR-20240902-hmjw" src="https://github.com/user-attachments/assets/e59e3317-9a72-4fca-a57e-2d00ac122e94">

```
root@gpustack:/# curl -v http://localhost/v1/workers
*   Trying 127.0.0.1:80...
* Connected to localhost (127.0.0.1) port 80 (#0)
> GET /v1/workers HTTP/1.1
> Host: localhost
> User-Agent: curl/7.81.0
> Accept: */*
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< date: Sun, 01 Sep 2024 21:55:11 GMT
< server: uvicorn
< content-length: 2207
< content-type: application/json
<
{"items":[{"name":"gpustack","hostname":"gpustack","ip":"172.22.0.42","labels":{},"system_reserved":{"memory":1073741824,"gpu_memory":1073741824},"state":"ready","state_message":null,"status":{"cpu":{"total":12,"utilization_rate":75.32894736842105},"memory":{"total":201079345152,"utilization_rate":13.99464841638913,"is_unified_memory":false,"used":28140347392,"allocated":0},"gpu_devices":[{"name":"AMD Device 164E (VGA compatible)","uuid":"","vendor":"AMD","index":0,"core":{"total":0,"utilization_rate":0.0},"memory":{"total":0,"utilization_rate":0.0,"is_unified_memory":false,"used":20168704,"allocated":null},"temperature":34.0},{"name":"NVIDIA GeForce RTX 3090","uuid":"GPU-875d550c-606a-9920-3296-7b219227b718","vendor":"NVIDIA","index":0,"core":{"total":10496,"utilization_rate":0.0},"memory":{"total":25335037952,"utilization_rate":0.016296671857458443,"is_unified_memory":false,"used":4128768,"allocated":null},"temperature":27.0},{"name":"NVIDIA RTX A4000","uuid":"GPU-389accfb-8815-6341-b1cc-82d6189527a0","vendor":"NVIDIA","index":2,"core":{"total":6144,"utilization_rate":0.0},"memory":{"total":16789602304,"utilization_rate":0.024591219763533956,"is_unified_memory":false,"used":4128768,"allocated":null},"temperature":31.0},{"name":"NVIDIA RTX A4000","uuid":"GPU-403d1335-aa67-7594-acbd-4b0095da1791","vendor":"NVIDIA","index":1,"core":{"total":6144,"utilization_rate":0.0},"memory":{"total":16789602304,"utilization_rate":0.024591219763533956,"is_unified_memory":false,"used":4128768,"allocated":null},"temperature":28.0}],"swap":{"total":47940997120,"utilization_rate":8.10147000964172,"used":3883925504},"filesystem":[{"name":"","mount_point":"/","mount_from":"overlay","total":2000398934016,"used":1389720170496,"free":610678763520,"available":604043972608}],"os":{"name":"Ubuntu","version":"22.04.4 LTS (Jammy Jellyfish)"},"kernel":{"name":"Linux","release":"6.10.6-200.fc40.x86_64","version":"#1 SMP PREEMPT_DYNAMIC Mon Aug 19 14:09:30 UTC 2024","architecture":""},"uptime":{"uptime":602417990.0,"boot_time":"2024-09-08T21:15* Connection #0 to host localhost left intact
:27.340+0000"}},"id":1,"created_at":"2024-08-06T23:22:36Z","updated_at":"2024-09-01T21:55:09Z"}],"pagination":{"page":1,"perPage":100,"total":1,"totalPage":1}}
```

**Comment by sammcj** [comment]
Then, when I add a model (ollama, name: llama3.1:8b-instruct-q6_K, model: llama3.1:8b-instruct-q6_K):

```
2024-09-01T22:00:31+00:00 - gpustack.server.controllers - DEBUG - Created model instance for model llama3.1:8b-instruct-q6_K
```

and it sits stuck in pending:

<img width="2487" alt="image" src="https://github.com/user-attachments/assets/a5b72670-c67f-4c80-b6ec-49ac994d5b7f">

This occurs even if I join other workers to the cluster, e.g:

<img width="2252" alt="image" src="https://github.com/user-attachments/assets/232e665b-8e75-4c59-a92f-741d1782a005">

<img width="2250" alt="image" src="https://github.com/user-attachments/assets/513e4a9d-5d35-4388-9b6e-ff84430880a2">

**Comment by sammcj** [comment]
Note that the `gpustack version` command doesn't seem to work with the default docker image:

```
gpustack version
0.0.0 (fb70b0e)
```

but looking at the commit history on the project - this is the latest commit.

**Comment by gitlawr** [comment]
It prints 0.0.0 because it's using `main(dev)` tag. We've updated it to the `latest` tag to avoid confusion.

We tried to reproduce, but no luck. Is it possible to install without docker in your environment? I’m just curious if the installation method might have anything to do with it.

**Comment by sammcj** [comment]
Yeah sure thing, I’ll give it a crack shortly tonight or tomorrow and report back.

Does the model deployment try to connect to a loopback address (that’s configurable), or does it use a socket / internal call?

**Comment by gitlawr** [comment]
Components like scheduler/controllers use internal calls.
The embedded worker that starts with the server connects to the loopback address. 
Other workers connects to the provided server address on start.

**Comment by sammcj** [comment]
I was able to get it working by disabling the embedded worker and explicitly providing the server URL, e.g:

```yaml
...
services:
  &name gpustack:
    container_name: *name
    hostname: *name
    profiles:
      - *name
    image: gpustack/gpustack:main
    ipc: host
    command: --debug --host 0.0.0.0 --data-dir /var/lib/gpustack --token mytokenhere --disable-worker
    ports:
      - 80
    volumes:
      - /mnt/llm/gpustack/data:/var/lib/gpustack
    networks:
      - traefik-servicenet

  &name gpustack-worker:
    <<: [*restart, *gpu]
    container_name: *name
    hostname: *name
    profiles:
      - gpustack
    image: gpustack/gpustack:main
    command: --server-url http://gpustack:80 --token mytokenhere
    links:
      - gpustack
    ports:
      - 80
    volumes:
      - /mnt/llm/gpustack/worker-data:/var/lib/gpustack
    networks:
      - traefik-servicenet
```

I suspect something in the app is trying to connect to or announce itself on one of the container network addresses rather than 127.0.0.1.

**Comment by gitlawr** [comment]
> something in the app
Is this referring to GPUStack or other services in the compose?
Could you provide more information about the Docker network so we can check if there's anything that needs to be resolved on our side?

**Comment by sammcj** [comment]
Sorry I was referring to within GPUStack, I was thinking perhaps there's some call to http://localhost rather than http://127.0.0.1 which can behave a little differently in containers.

**Comment by gitlawr** [comment]
It explicitly uses the loopback address(https://github.com/gpustack/gpustack/blob/4d42dce2fd12b88ac80a33001c210687d53ea8bc/gpustack/cmd/start.py#L203). I tried modifying the hosts file so that localhost resolves to something else, and it works normally, so the resolution of localhost might not be the cause.

**Comment by gitlawr** [comment]
We identified and fixed [issue #284](https://github.com/gpustack/gpustack/issues/284) in version 0.2.1. Hopefully, this fix should also resolve the issue you're experiencing. If it doesn't, the logs should provide more detailed information. 

Feel free to provide feedback or reopen the issue if you have any further questions or encounter any problems.

**Comment by sammcj** [comment]
Thank you for looking into this Lawrence, I really appreciate it.
